File: RxTango/python/examples/read_attribute.py
"""Read one Tango attribute — single-shot Observable.

Equivalent to `caget` but as a reactive stream: emits one value and completes.

Usage:
    python read_attribute.py [device] [attribute]

    device     defaults to tango://localhost:10000/sys/tg_test/1
    attribute  defaults to double_scalar
"""

import asyncio
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parent.parent / "src"))

from reactivex.scheduler.eventloop import AsyncIOScheduler
from rxtango import read_attribute


async def main() -> None:
    device = sys.argv[1] if len(sys.argv) > 1 else "tango://localhost:10000/sys/tg_test/1"
    attr   = sys.argv[2] if len(sys.argv) > 2 else "double_scalar"

    loop      = asyncio.get_running_loop()
    scheduler = AsyncIOScheduler(loop)
    done      = asyncio.Event()

    print(f"Reading {device}/{attr} …")

    read_attribute(device, attr).subscribe(
        on_next=lambda v: print(f"  value = {v}"),
        on_error=lambda e: (print(f"  ERROR: {e}", file=sys.stderr), done.set()),
        on_completed=done.set,
        scheduler=scheduler,
    )

    await done.wait()


if __name__ == "__main__":
    asyncio.run(main())

File: RxTango/cpp/examples/read_attribute.cpp
/**
 * Single-shot attribute read — the simplest possible RxTango example.
 *
 * Mirrors Python's read_attribute.py and Java's ReadAttribute.java.
 *
 * Usage:
 *   ./read_attribute [device] [attribute]
 *   defaults: tango://localhost:10000/sys/tg_test/1  double_scalar
 */

#include <condition_variable>
#include <iostream>
#include <mutex>
#include <string>

#include <rxtango/rxtango.hpp>

int main(int argc, char* argv[]) {
    const std::string device = argc > 1 ? argv[1] : "tango://localhost:10000/sys/tg_test/1";
    const std::string attr   = argc > 2 ? argv[2] : "double_scalar";

    std::cout << "Reading " << device << "/" << attr << " ...\n";

    std::mutex              m;
    std::condition_variable cv;
    bool                    done = false;

    rxtango::read_attribute<double>(device, attr).subscribe(
        [](double v) { std::cout << "  value: " << v << "\n"; },
        [&](std::exception_ptr e) {
            try { std::rethrow_exception(e); }
            catch (std::exception& ex) { std::cerr << "  ERROR: " << ex.what() << "\n"; }
            std::lock_guard<std::mutex> lk(m); done = true; cv.notify_one();
        },
        [&]() {
            std::lock_guard<std::mutex> lk(m); done = true; cv.notify_one();
        }
    );

    std::unique_lock<std::mutex> lk(m);
    cv.wait(lk, [&]{ return done; });
    return 0;
}
